src/ray/raylet/local_object_manager.cc

```diff
@@ -347,7 +352,9 @@ void LocalObjectManager::AsyncRestoreSpilledObject(
     const ObjectID &object_id, const std::string &object_url, const NodeID &node_id,
     std::function<void(const ray::Status &)> callback) {
   if (objects_pending_restore_.count(object_id) > 0) {
-    // If the same object is restoring, we dedup here.
+    // If the same object is restoring, we dedup here. We don't invoke the callback in
```

Comment:
That's a good point, this should never be reached AFAICT.

Reply:
Hmm actually @rkooo567, I hadn't thought of this, but I don't think that we're handling concurrent restoration requests from different nodes for the same object correctly. Consider this case (which I think can happen):

1. Node 1 spills object.
2. Node 2 issues remote restoration request to node 1.
3. Concurrently, node 1 issues a local restoration request.
4. The callbacks for one of these restoration requests will not be invoked due to deduping (with the current implementation), causing one of these pulls to never succeed.

With your suggestion (adding a RAY_CHECK here), one of those restoration requests would trip that check and cause a crash.

Am I right that this can happen? If so, we'd probably want restoration requests from different nodes to join on a single restoration that's processed, where each callback is invoked once the restoration is done, right?

Reply:
Followed up on slack thread.